maestro/trainer/models/florence_2/metrics.py

```diff
@@ -0,0 +1,103 @@
+import re
+from typing import List, Dict
+from typing import Tuple
+
+import numpy as np
+import supervision as sv
+import torch
+from PIL import Image
+from supervision.metrics.mean_average_precision import MeanAveragePrecision
+from tqdm import tqdm
+from transformers import AutoProcessor, AutoModelForCausalLM
+
+from maestro.trainer.common.data_loaders.datasets import DetectionDataset
+from maestro.trainer.common.utils.metrics import BaseMetric
+
+DETECTION_CLASS_PATTERN = r"([a-zA-Z0-9 -]+)<loc_\d+>"
+
+
+class MeanAveragePrecisionMetric(BaseMetric):
+
+    def describe(self) -> List[str]:
+        return ["map50:95", "map50", "map75"]
+
+    def compute(
+        self,
+        targets: List[sv.Detections],
+        predictions: List[sv.Detections]
+    ) -> Dict[str, float]:
+        result = MeanAveragePrecision().update(targets=targets, predictions=predictions).compute()
+        return {
+            "map50:95": result.map50_95,
+            "map50": result.map50,
+            "map75": result.map75
+        }
+
+
+def postprocess_florence2_output_for_mean_average_precision(
+    expected_responses: List[str],
+    generated_texts: List[str],
+    images: List[Image.Image],
+    classes: List[str],
+    processor: AutoProcessor
+) -> Tuple[List[sv.Detections], List[sv.Detections]]:
+    targets = []
+    predictions = []
+    
+    for image, suffix, generated_text in zip(images, expected_responses, generated_texts):
+        # Postprocess prediction for mean average precision calculation
+        prediction = processor.post_process_generation(generated_text, task="<OD>", image_size=image.size)
+        prediction = sv.Detections.from_lmm(sv.LMM.FLORENCE_2, prediction, resolution_wh=image.size)
+        prediction = prediction[np.isin(prediction["class_name"], classes)]
```

Comment:
I also agree that this is one option to choose `text-based output` to calculate traditional OD.
However, some prediction is very close to the `classes`, but this `np.isin` will not able to catch.
e.g. prediction : apple, groundtruth : apples.

I also considered to use calculate the distance vectorized text embedding or other heuristic method such as `CIDEr` for to make more robust. It would be great to consider metrics in VLM. e.g. `CIDEr`, `BLUE`, etc...

Reply:
> I also agree that this is one option to choose `text-based output` to calculate traditional OD.
However, some prediction is very close to the `classes`, but this `np.isin` will not able to catch.
e.g. prediction : apple, groundtruth : apples.

Good catch! I've experienced that myself. I don't have the time to address it right now, but I'll add a task for it. Maybe one of the external contributors would like to implement this feature.chciałby zaimplementować ten feature.

Reply:
https://huggingface.co/metrics

Reply:
https://www.cv-foundation.org/openaccess/content_cvpr_2015/papers/Vedantam_CIDEr_Consensus-Based_Image_2015_CVPR_paper.pdf